plugins/console/index.js

```diff
@@ -0,0 +1,127 @@
+ *
+ *    exports.config = {
+ *      plugins: [{
+ *        path: 'node_modules/protractor/plugins/console',
+ *        failOnWarning: {Boolean}                (Default - false),
+ *        failOnError: {Boolean}                  (Default - true)
+ *        exclude: {Array of strings and regex}   (Default - [])
+ *      }]
+ *    };
+ */
+var ConsolePlugin = function() {
+};
+
+/**
+ * Gets the browser log
+ *
+ * @returns {!webdriver.promise.Promise.<!Array.<!webdriver.logging.Entry>>}
+ */
+ConsolePlugin.getBrowserLog = function() {
+  return browser.manage().logs().get('browser');
+};
+
+/**
+ * Logs messages to the test output
+ *
+ * @param warnings
+ *      The list of warnings detected by the browser log
+ * @param errors
+ *      The list of errors detected by the browser log
+ */
+ConsolePlugin.logMessages = function(warnings, errors) {
+  warnings.map(function(warning) {
+    testOut.specResults.push({
+      description: warning.level.name,
+      passed: false,
+      assertions: [{passed: false, errorMsg: warning.message}]
+    });
+  });
+
+  errors.map(function(error) {
+    testOut.specResults.push({
+      description: error.level.name,
+      passed: false,
+      assertions: [{passed: false, errorMsg: error.message}]
+    });
+  });
+};
+
+/**
+ * Determines if a log message is filtered out or not. This can be set at the config stage using the exclude parameter.
+ * The parameter accepts both strings and regex
+ *
+ * @param logMessage
+ *      Current log message
+ * @returns {boolean}
+ */
+ConsolePlugin.includeLog = function(logMessage) {
+  return this.exclude.filter(function(e) {
+      return (e instanceof RegExp) ? logMessage.match(e) : logMessage.indexOf(e) > -1;
+    }).length === 0;
+};
+
+/**
+ * Parses the log and decides whether to throw an error or not
+ *
+ * @param config
+ *      The config from the protractor config file
+ * @returns {Deferred.promise}
+ */
+ConsolePlugin.parseLog = function(config) {
+  var self = this;
+  var deferred = q.defer();
+  var failOnWarning = config.failOnWarning || false;
+  var failOnError = config.failOnError || true;
+  this.exclude = config.exclude || [];
+
+  this.getBrowserLog().then(function(log) {
+
+    var warnings = log.filter(function(node) {
+      return (node.level || {}).name === 'WARNING' && self.includeLog(node.message);
+    });
+
+    var errors = log.filter(function(node) {
+      return (node.level || {}).name === 'SEVERE' && self.includeLog(node.message);
+    });
+
+    testOut.failedCount += (warnings.length > 0 && failOnWarning) ? 1 : 0;
+    testOut.failedCount += (errors.length > 0 && failOnError) ? 1 : 0;
+
+    if(testOut.failedCount > 0) {
+      self.logMessages(warnings, errors);
+    }
+
+    deferred.resolve();
+  });
+
+  return deferred.promise;
+};
+
+/**
+ * Tear-down function used by protractor
+ *
+ * @param config
+ */
+ConsolePlugin.prototype.teardown = function(config) {
+  var audits = [];
```

Comment:
Hi Sammy

I had initially set it up like this but ran into a strange issue where the test would finish before the parseLog function finished which meant testOut.failedCount was always zero and it would never fail the test. The only way I could get it to work in order was to wrap it in a promise